blockstore/apps/bundles/links.py

```diff
@@ -0,0 +1,231 @@
+"""
+Classes for storing Link related information in our Draft and Snapshot repos.
+
+TODO: store.py should probably be split into its own package with this as one of
+the modules in that package.
+"""
+from typing import List
+from uuid import UUID
+import codecs
+
+import attr
+
+
+@attr.s(frozen=True)
+class Dependency:
+    """
+    A Dependency is a pointer to exactly one Bundle + Version + Snapshot.
+    """
+    bundle_uuid = attr.ib()
+    version = attr.ib(type=int)
+    snapshot_digest = attr.ib(type=bytes)
+
+    @classmethod
+    def from_json_dict(cls, json_dict):
+        """Parse Dependency from a dict."""
+        return cls(
+            bundle_uuid=UUID(json_dict["bundle_uuid"]),
+            version=json_dict["version"],
+            snapshot_digest=bytes_from_hex_str(json_dict["snapshot_digest"]),
+        )
+
+
+@attr.s(frozen=True)
+class Link:
+    """
+    A Link is a named Dependency + all of its transitive dependencies.
+
+    Links are named because it's totally fine to have Links that point to
+    different versions of a Bundle (e.g Bundle Av1, Bundle Av2), but we want
+    some stable way of identifying and differentiating between those references.
+
+    The indirect dependencies are a flat list. We don't try to keep track of all
+    the nested dependency relations. Instead, we want the Link to answer the
+    question of "If I wanted to download this entire Bundle and all the things
+    it links to, what is the complete list of other BundleVersions that I have
+    to download to get all the dependencies?
```

Comment:
Will this be a problem for OLX parsing clients?

i.e. if we end up with a bundle that has an OLX file which says:

```
<xblock-include definition="linked_bundle1/dep2/dep3/dep4/dep5/dep6/dep7/html/intro" usage="blah" />
```
(hopefully that is an extremely unusual example, or perhaps even forbidden; but nonetheless: )

Would the client then be expected to make 7+ API calls to load the link information for those bundles one at a time in order to resolve the name of each component in the "path" to a BundleVersion, to be able to identify which BundleVersion contains the referenced file?

That may be a reasonable trade-off, just want to clarify.

Reply:
This is a client decision, but I don't the Studio client here should allow direct linking to things that are dependencies of dependencies. If I have a piece of content that I depend on explicitly, then I should add that as a named Link to my Bundle, and make that reference directly. Does that seem reasonable?

Reply:
Yep, that seems reasonable.